File: ios/Sources/RevenueDog/Models/Package.swift
import Foundation

/// The semantic duration class of a ``Package``.
public enum PackageType: Sendable, Equatable {
    case lifetime
    case annual
    case sixMonth
    case threeMonth
    case twoMonth
    case monthly
    case weekly
    /// A package with a non-standard identifier.
    case custom
    /// A package whose identifier could not be classified.
    case unknown

    /// Derives the package type from a RevenueDog package identifier.
    ///
    /// Standard identifiers map to their corresponding case; any other
    /// `$`-prefixed identifier maps to ``custom``.
    init(identifier: String) {
        switch identifier {
        case SDKConstants.lifetimePackageID:   self = .lifetime
        case SDKConstants.annualPackageID:     self = .annual
        case SDKConstants.sixMonthPackageID:   self = .sixMonth
        case SDKConstants.threeMonthPackageID: self = .threeMonth
        case SDKConstants.twoMonthPackageID:   self = .twoMonth
        case SDKConstants.monthlyPackageID:    self = .monthly
        case SDKConstants.weeklyPackageID:     self = .weekly
        default:                               self = .custom
        }
    }
}

/// A purchasable bundle pairing a RevenueDog package identifier with a
/// concrete, store-resolved ``StoreProduct``.
public struct Package: Sendable {
    /// The RevenueDog package identifier (e.g. `"$rd_monthly"`).
    public let identifier: String
    /// The semantic duration class of the package.
    public let packageType: PackageType
    /// The fully-resolved store product (already fetched from the App Store).
    public let storeProduct: StoreProduct
    /// The identifier of the offering this package belongs to.
    public let offeringIdentifier: String

    public init(identifier: String,
                packageType: PackageType,
                storeProduct: StoreProduct,
                offeringIdentifier: String) {
        self.identifier = identifier
        self.packageType = packageType
        self.storeProduct = storeProduct
        self.offeringIdentifier = offeringIdentifier
    }
}

File: android/revenuedog/src/main/kotlin/com/revenuedog/purchases/models/Package.kt
package com.revenuedog.purchases.models

/** Duration category of a [Package], derived from its `$rd_*` identifier. */
enum class PackageType(val identifier: String?) {
    LIFETIME("\$rd_lifetime"),
    ANNUAL("\$rd_annual"),
    SIX_MONTH("\$rd_six_month"),
    THREE_MONTH("\$rd_three_month"),
    TWO_MONTH("\$rd_two_month"),
    MONTHLY("\$rd_monthly"),
    WEEKLY("\$rd_weekly"),

    /** A non-reserved identifier chosen by the developer. */
    CUSTOM(null),

    /** A reserved (`$`-prefixed) identifier this SDK version does not know. */
    UNKNOWN(null);

    companion object {
        /** Maps a package identifier to its [PackageType]. */
        fun fromIdentifier(identifier: String): PackageType =
            entries.firstOrNull { it.identifier == identifier }
                ?: if (identifier.startsWith("$")) UNKNOWN else CUSTOM
    }
}

/**
 * A purchasable entry of an [Offering], pairing a package identifier (e.g. `$rd_monthly`)
 * with the [StoreProduct] already fetched from Google Play.
 */
class Package internal constructor(
    val identifier: String,
    val packageType: PackageType,
    val storeProduct: StoreProduct,
    /** Identifier of the offering this package was presented in. */
    val offeringIdentifier: String
) {
    override fun toString(): String =
        "Package(identifier=$identifier, product=${storeProduct.productIdentifier})"
}
